### ISSUE TITLE
Support Java 8 type annotations

### ISSUE THREAD

**Issue by jbrower2** [classgraph/classgraph#402]
In the Classfile class, the attributes "RuntimeVisibleTypeAnnotations" and "RuntimeInvisibleTypeAnnotations" are not handled. These attributes can be present on classes, methods, and fields. See [the class file format spec](https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.7-320).

**Comment by lukehutch** [comment]
That link seems broken right now, but here's the same documentation in the JDK 13 docs:

https://docs.oracle.com/javase/specs/jvms/se13/html/jvms-4.html#jvms-4.7.20

Do you have any interest in giving it a shot?

**Comment by jbrower2** [comment]
@lukehutch I took an initial pass at it, but it seems more involved than I was expecting. I'm not familiar enough with the project structure to dive that deep. Sorry!

**Comment by lukehutch** [comment]
No problem, thanks for taking a look. How urgently do you need type annotation support?

Also, I'm curious, what are you using type annotations for?

**Comment by jbrower2** [comment]
It would be great to have support immediately, but I understand how much development this would be, so I won't rush you.

We are writing a maven plugin that analyzes [bean validation annotations](https://www.baeldung.com/javax-validation) model classes and generates [json-schema](https://json-schema.org/) for validating API requests/responses. Most of our use cases are covered with your current implementation, for example:
```
public class Person {
  @NotNull
  @Size(max = 30)
  private String name;
}
```
will generate a json-schema:
```
{
  "Person": {
    "properties": {
      "name": {
        "type": "string",
        "maxLength": 30
      }
    },
    "required": [ "name" ]
  }
}
```

But we are not able to capture type annotations:
```
public class Person {
  private List<@Size(max = 50) String> emails;
}
```
(this would specify that each email address of this person can be 50 characters at most)

So it's a pretty rare edge case, but we do have examples of it happening.

This validation we're doing is between layers of our application, and the API layer will validate the input itself, but the error messaging is a bit harder to maintain at that layer, so we're trying to avoid it.

Hopefully some of this makes sense. I really appreciate the fast turnaround! :)

**Comment by jbrower2** [comment]
Also a side note, I do have a good amount of experience with the ASM library and have thought about migrating my plugin to use that, but it would be a pretty significant overhaul (not my first rewrite of it, though...).

**Comment by lukehutch** [comment]
That's a really useful piece of software that you're developing, and I hope you can make a standalone library and not just a Maven plugin, because projects like the [serverx](https://github.com/lukehutch/serverx) library that I wrote could use this to ensure that HTML elements in forms are bound to POJOs on the server with exactly the same validation logic on both server and client side.

A word of caution on ASM: ASM stops with an exception when you try to parse classfiles newer than the newest stable release at the time the ASM library version was cut -- and the JDK team has decided to bump the classfile version number every time a new JDK release comes out, i.e. every 6 months, even if older classfile parsing libraries could continue to parse newer classfiles without any problems. This didn't use to be a problem, since JDK releases were so rare, but with the new cadence, it means that if you depend upon ASM, you have to push a new release of your library out every 6 months, every time the new ASM for the latest JDK is released -- or your library will start bombing on some classfiles. 

ClassGraph on the other hand takes the view that because Java always tries to maintain backwards compatibility, typically changes to the classfile format should be backwards compatible (e.g. attributes can be added to methods/fields/classes, but those can be ignored by older tools if they're not understood, e.g. "nest mate" information). I found one place in the classfile format where this clearly wasn't the case, which was when new constant pool types were added to the constant pool (it happened in JDK 9 and again in JDK 11 or so) -- constant pool entries do not contain a size field, so you can't just skip them if you don't know how to interpret them. I added my thoughts on this to a thread someone else started on jdk-dev entitled [Please stop incrementing the classfile version number when there are no format changes](https://mail.openjdk.java.net/pipermail/jdk-dev/2019-October/003388.html), and the all-out assault by Brian Goetz on being able to rely on any sort of backwards compatibility in the classfile format was extremely disappointing. The ASM author also contributed to that conversation and basically said that if that's JDK's policy, we'll keep pushing out a release every 6 months, and all our downstream dependents will have to upgrade every 6 months too.

So move to ASM if you want, but be aware of this caveat! ClassGraph will continue to ignore the classfile version code until or unless something breaks (usually by the addition of a new constant pool type, which should in general happen only very rarely).

**Comment by lukehutch** [comment]
PS the type annotation system is extremely complex, based on the documentation, and the worst part is that it's a "bolt-on" (again in the name of backwards compatibility), so you have to carefully correlate type references with annotation according to indices in the type annotation data. This will be hard to get right, so it'll have done carefully and properly unit tested, otherwise annotations and types could get out of sync. It will require changes to the `Classfile` class for parsing, and also to all the type representation classes. Type parsing is lazy though for speed, so annotating types with type annotations will also need to be done lazily. It's going to require some work to get this right. It looks worth doing though, so I'll try to give it a shot sometime soon.

**Comment by GedMarc** [comment]
those are javax/jakarta validation annotations not jdk 8 ones?

you will need the jakarta validation library included in your scanresult - I have confirmed that the annotations are found under these conditions.

**Comment by lukehutch** [comment]
@jbrower2 Sorry for the delay, I have been swamped for months with other projects, so wasn't able to prioritize feature requests for a while.

I implemented support for Java type annotations, and it's checked into git. This was extremely fiddly to get right...

I implemented some basic unit tests, using the examples shown in **4.7.20.2. The type_path structure** of the Java spec (section numbering from the Java 13 doc link I posted above):

https://github.com/classgraph/classgraph/blob/latest/src/test/java/io/github/classgraph/issues/issue402/TypeAnnotationTest.java

This just ensures that the `toString()` method has the annotations in the correct place.

You will need to call `.enableAnnotationInfo()` or `.enableAllInfo()` in your code before `.scan()`. You can read type annotations by calling e.g. `ClassInfo#getTypeDescriptorOrTypeSignature()`, followed by calling the `getTypeAnnotationInfo()` method to get an `AnnotationInfoList` of type annotations.

If you're dealing with `ClassRefTypeSignature`, you also have a `getSuffixTypeAnnotationInfo()` for nested annotations on type suffixes, e.g. for the first parameter of type `List<@A X.@B Y.@C Z>`, `getTypeAnnotationInfo` will return `[@A]`, whereas `getSuffixTypeAnnotationInfo()` will return `[[@B], [@C]]`, since `getBaseClassName()` returns `"X"` and `getSuffixes()` returns `["Y", "Z"]`.

**Comment by lukehutch** [comment]
OK, I think most of the bugs are fixed now...

**Comment by lukehutch** [comment]
I'm pretty sure this is complete and working. Released in 4.8.96.

Please feel free to add more unit tests, and please open another bug report if you find any problems.

Thanks for the feature request!
